wiki/DEBUGGING.md

```diff
@@ -0,0 +1,128 @@
+a specific function in the module or `:break ModuleA line_number` if you want to set the breakpoint at a specific line.
+In order to see the breakpoints set, use the command `:show breaks`. Breakpoints can be enabled/disabled with the
+command `:enable/:disable breakpoint_number`. To enable/disable all breakpoints at the same time use
+`:enable/:disable *`. Finally, to delete breakpoints use the command `:delete breakpoint_number` or to delete all use
+`:delete *`.
+
+### Program Trace
+
+Once you have your breakpoints set, you can start running the program through the main entrypoint `main`. After, the
+debugger will stop wherever you set your breakpoint. From this point the relevant variables and functions the program
+is using should be viewable. Use the command `:info var_name` to see the object's type and `:print var_name` to see its
+value. **Note that you can only use the main entry point once, i.e. you will have to restart the debugger**.
+
+Stepping through the program can be done using `:step`. This will step through the program until the next breakpoint.
+There are two variants `:steplocal` and `:stepmodule`, where the former limits the breakpoints that are enabled to the
+scope of the current function and the latter restricts the scope of the breakpoints to the current module. **To reiterate
+the point about the main entry point in the last paragraph, the first time `main` is called it will trace step by step
+until the result of the entire program is printed, but if main is called again in the same session, it will be evaluated
+and the stepper environment will not be available**.
+
+Functions can also be called. Type the function and its input parameters directly into the debugger: `function inputs`.
+The downside about doing this is that information is lost with respect to specific input values if this is done during
+debugging with breakpoints. To be more specific, the debugger will tell you the algebraic types of the variables,
+but it won't tell you what the actual values are.
+
+You can also see an entire trace of debugging a function with `:trace`. Set the breakpoint at the function's base case/
+step before evaluation. Then use `:hist` to see the previous steps. Use the command `:back` to backtrack steps. This
+can be useful if you want to revisit one of the steps during the debug process, especially in recursive functions.
+
+### Example
+
+Here we use a toy project with modules `Main` and `Functions.Extra1`:
+
+Main:
+
+```aiignore
+import Functions.Extra1(add2, sumList, foldList)
+
+main :: IO ()
+main = print (add2 1 2 + sumList [3, 4, 5, 6] + foldList [7, 8, 9, 10])
+```
+
+Functions.Extra1:
+
+```aiignore
+module Functions.Extra1 where
+
+add2 ::(Num a) => a -> a -> a
+add2 x y = x + y
+
+sumList :: (Num a) => [a] -> a
+sumList []     = 0
+sumList (x:xs) = x + sumList xs
+
+foldList :: (Num a) => [a] -> a
+foldList x = foldl (+) 1 x
```

Comment:
Improve both the function name and parameter name. Also don't provide the buggy version here, this isn't necessary for illustrating the debugger.